### PR TITLE
hwt: Rework buffer notification mechanism + various fixes

### PR DESCRIPTION
**Pull request by bnovkov**
This PR reworks hwt's buffer notification mechanism and fixes various panics/bugs that were encountered during testing.

Here's a rough list of changes:
- Reworked `struct hwt_record_entry` into a tagged union with helper functions
- Added new record type - `HWT_RECORD_BUFFER`
- Added ability to block in `hwt_record_fetch`
- Replaced `hwt_event` with `HWT_RECORD_BUFFER` + blocking `hwt_record_fetch`
- Added dedicated UMA zone for allocating hwt records
- Changed PT driver to use `taskqueue` + `HWT_RECORD_BUFFER`
- Added new userspace backend operations along with a default data processing loop that uses blocking `hwt_record_fetch` to service buffers
- Fix for a crash caused by an improper free of backend-specific thread data
- Fixes for issues found by `clang-tidy` in hwt_pt.c 
- Several driver and userspace fixes for PT CPU mode

### CODE REVIEW

File: usr.sbin/hwt/hwt.c [hunk 1]
```diff
@@ -318,6 +318,72 @@ usage(void)
...
+			error = hwt_record_fetch(tc, &nrec, 1);
+			if (error != 0 || nrec == 0)
+				break;
+		}
+		if (errno == EINTR || tc->terminate) {
+			printf("%s: tracing terminated - exiting\n", __func__);
+			/* Fetch any remaining records */
+			hwt_record_fetch(tc, &nrec, 0);
+			if (tc->trace_dev->methods->shutdown != NULL)
+				;
```
Comment (bukinr) [inline usr.sbin/hwt/hwt.c]: typo ?

File: usr.sbin/hwt/hwt_record.c [hunk 2]
```diff
@@ -167,7 +174,15 @@ hwt_record_fetch(struct trace_context *tc, int *nrecords)
 				return (error);
 			break;
 		case HWT_RECORD_THREAD_SET_NAME:
+		case HWT_RECORD_MUNMAP:
 			break;
+		case HWT_RECORD_BUFFER:
+			/* Invoke backend process method. */
+			assert(tc->trace_dev->methods->process_buffer != NULL);
+			if ((error = tc->trace_dev->methods->process_buffer(tc,
+				 entry->buf_id, entry->curpage,
```
Comment (bukinr) [inline usr.sbin/hwt/hwt_record.c]: extra space character before entry->

File: usr.sbin/hwt/hwt_record.c [hunk 3]
```diff
@@ -147,16 +151,19 @@ hwt_record_fetch(struct trace_context *tc, int *nrecords)
...
-		case HWT_RECORD_MUNMAP:
 		case HWT_RECORD_EXECUTABLE:
 		case HWT_RECORD_INTERP:
 		case HWT_RECORD_KERNEL:
 			hwt_record_to_elf_img(tc, entry, &img);
 			/* Invoke backend callback, if any. */
 			if (tc->trace_dev->methods->image_load_cb != NULL &&
 			    (error = tc->trace_dev->methods->image_load_cb(tc,
-					 &img) != 0))
+				 &img))) {
```
Comment (bukinr) [inline usr.sbin/hwt/hwt_record.c]: extra space character before &img

File: usr.sbin/hwt/hwt.c (hunk 1 above)
Comment (bnovkov) [reply usr.sbin/hwt/hwt.c]: thanks for catching this!
Fixed.

**Pull request merged**